### PR TITLE
feat: Initial implementation of RePO trainer and components

### PR DESCRIPTION
**Pull request by celsowm**
This commit introduces the initial framework for the Replay-Enhanced Policy Optimization (RePO) trainer within the TRL library. (https://github.com/SihengLi99/RePO)

Key changes include:

1.  **RePOConfig:** Added `trl/trainer/repo_config.py` defining `RePOConfig`, which inherits from `PPOConfig` and includes RePO-specific hyperparameters such as `replay_buffer_size`, `replay_batch_size`, `kl_coef_replay`, `importance_sampling_clip`, `replay_warmup_steps`, and `use_replay_buffer`. Updated `trl/trainer/__init__.py` to expose `RePOConfig`.

2.  **ReplayBuffer:** Implemented a `ReplayBuffer` class in `trl/trainer/utils.py`. This buffer stores experiences (dictionaries of tensors) on the CPU and provides methods to `add` new experiences and `sample` batches, moving them to a specified device. Updated `trl/trainer/__init__.py` to expose `ReplayBuffer`.

3.  **RePOTrainer - Initial Structure:** Created `trl/trainer/repo_trainer.py` with the `RePOTrainer` class, inheriting from `PPOTrainer`.
    - The `__init__` method initializes `RePOConfig` and the `ReplayBuffer`.
    - The `train` method is inherited from `PPOTrainer` for now.
    - The `compute_loss` method has been overridden to include:
        - Calculation of the on-policy PPO loss by calling `super().compute_loss()`.
        - Addition of on-policy experiences (ppo_elements) to the replay buffer. - Logic for sampling from the replay buffer after warmup steps. - Recomputation of log probabilities for off-policy data using the current policy. - Calculation of importance sampling weights (rho) with clipping. - An initial implementation of the off-policy policy loss component based on importance-weighted advantages from the buffer. - An initial implementation of KL regularization for off-policy data against the reference model, scaled by `kl_coef_replay`. - Combination of on-policy and off-policy loss components. - Addition of relevant metrics (buffer size, rho, off-policy losses, KL divergence) to the metrics dictionary.
    - Corrected import for log probability calculation to use `selective_log_softmax` from `trl.trainer.utils` instead of the non-existent `logprobs_from_logits` from `trl.core`.
    - Updated `trl/trainer/__init__.py` to expose `RePOTrainer`.

4.  **Basic Tests:** Added `tests/trainer/test_repo_trainer.py` with:
    - Unit tests for `ReplayBuffer` (add, sample, capacity, device movement).
    - A basic initialization test for `RePOTrainer`.
    - A simple run-through test for `RePOTrainer` to ensure a few training steps (on-policy with buffer filling and initial off-policy logic) can execute without crashing using a small model and dataset.

**Next Steps/Current Status:**
The core structure for RePO is in place. The `compute_loss` method contains the initial logic for both on-policy and off-policy updates. Further work would involve:
- Rigorous testing of the off-policy loss calculations and their impact.
- Potentially overriding `PPOTrainer.ppo_train_loop` or `stat_tracking_inner_loop` for more fine-grained control over when experiences are added to the replay buffer if the current approach within `compute_loss` proves problematic with the `inputs` structure.
- More comprehensive testing, including checking the numerical correctness of loss components and the behavior of the replay mechanism under various conditions.
- Performance optimization.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by shirinyamani** [comment]
Nice work @celsowm!
can you also please provide a minimal script so that we can run and test it?

**Comment by celsowm** [comment]
> Nice work @celsowm! can you also please provide a minimal script so that we can run and test it?

```python
import torch
from datasets import load_dataset
from peft import LoraConfig
from transformers import AutoTokenizer, AutoModelForCausalLMWithValueHead, BitsAndBytesConfig
from trl import RePOConfig, RePOTrainer

repo_config = RePOConfig(
    model_name="Qwen/Qwen2-0.5B-Instruct",
    learning_rate=1.4e-5,
    batch_size=4,
    ppo_epochs=2,
    log_with="wandb", 
    replay_buffer_size=500,
    replay_batch_size=4,
    replay_warmup_steps=10,
)

quantization_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16,
)

peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=[
        "q_proj", "k_proj", "v_proj", "o_proj",
        "gate_proj", "up_proj", "down_proj",
    ],
)

model = AutoModelForCausalLMWithValueHead.from_pretrained(
    repo_config.model_name,
    quantization_config=quantization_config,
    device_map="auto",
    trust_remote_code=True,
    peft_config=peft_config,
)

tokenizer = AutoTokenizer.from_pretrained(
    repo_config.model_name,
    trust_remote_code=True,
)
tokenizer.pad_token = tokenizer.eos_token

dataset = load_dataset("lvwerra/stack-exchange-paired", split="train[:1%]")

def build_dataset(tokenizer, dataset):

    def format_with_template(sample):
        messages = [{"role": "user", "content": f"Question: {sample['question']}\n\nAnswer:"}]
        sample["query"] = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
        return sample

    dataset = dataset.map(format_with_template)
    tokenized_dataset = dataset.map(lambda x: tokenizer(x["query"], truncation=True, max_length=128), batched=True)
    tokenized_dataset.set_format(type="torch")
    return tokenized_dataset

processed_dataset = build_dataset(tokenizer, dataset)

repo_trainer = RePOTrainer(
    config=repo_config,
    model=model,
    ref_model=None, # Automatically created by the trainer
    tokenizer=tokenizer,
    dataset=processed_dataset,
)

generation_kwargs = {
    "min_length": -1,
    "top_k": 0.0,
    "top_p": 1.0,
    "do_sample": True,
    "pad_token_id": tokenizer.eos_token_id,
    "max_new_tokens": 64,
}

print("Starting RePO training with the official TRL implementation...")
stats = repo_trainer.train(generation_kwargs=generation_kwargs)
print("RePO training finished successfully.")
print("Final Stats:", stats)
```

**Comment by qgallouedec** [comment]
Thanks for this @celsowm, and sorry for the long delay getting back to you.

Since this was opened, PPO was removed from TRL core and moved to experimental (#5051), so RePOConfig(PPOConfig) / RePOTrainer(PPOTrainer) no longer import against main. On top of that, the RePO paper actually builds on GRPO rather than PPO, and the PR itself is described as an initial WIP with the core off-policy logic still to be validated. Furthermore, it has received limited attention since its publication.

Given that plus ~12 months of inactivity, I'm going to close this. Really appreciate the effort here though!
